### PR TITLE
Side panels resizing

### PR DESCRIPTION
**Pull request by lorem--ipsum**

### CODE REVIEW

File: src/client/utils/local-storage/local-storage.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+export function supported(): boolean {
+  return isSupported;
+}
+
+export function get(key: string): any {
+  if (!isSupported) {
+    throw new Error('LocalStorage not supported');
+  }
```
Comment (vogievetsky) [inline src/client/utils/local-storage/local-storage.ts]: It is crazy for get to throw an error if localstorage is not supported. Instead you should pass a second argument for the default value and that should be (quietly) used.

Comment (vogievetsky) [inline src/client/utils/local-storage/local-storage.ts]: Or just return undefined...

File: src/client/utils/local-storage/local-storage.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+var isSupported = false;
+
+var mod = 'test';
+try {
+    localStorage.setItem(mod, mod);
+    localStorage.removeItem(mod);
+    isSupported = true;
+} catch (e) {
+    isSupported = false;
+}
```
Comment (vogievetsky) [inline src/client/utils/local-storage/local-storage.ts]: is this necessary? Why not just wrap every local storage call in a try catch?

Comment (lorem--ipsum) [inline src/client/utils/local-storage/local-storage.ts]: I can do that :)

File: src/client/components/cube-view/cube-view.tsx [hunk 3]
```diff
@@ -390,7 +448,18 @@ export class CubeView extends React.Component<CubeViewProps, CubeViewState> {
...
           </div>
         </div>
+
+        <ResizeHandle
+          side="right"
+          initialValue={layout.right}
+          onResize={this.onRightPanelResize.bind(this)}
+          onResizeEnd={this.onPanelResizeEnd.bind(this)}
+          min={240}
+          max={400}
```
Comment (vogievetsky) [inline src/client/components/cube-view/cube-view.tsx]: Can you extract these into constants at the top of the file please

File: src/client/components/resize-handle/resize-handle.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  getValueFromX(x: number): number {
+    if (this.props.side !== 'right') {
+      return this.constrainValue(x);
+    }
+
+    return this.constrainValue(window.innerWidth - x);
+  }
+
+  constrainValue(value: number): number {
+    return Math.max(Math.min(value, this.props.max), this.props.min);
```
Comment (vogievetsky) [inline src/client/components/resize-handle/resize-handle.tsx]: search for "clamp" in the source code

Comment (vogievetsky) [inline src/client/components/resize-handle/resize-handle.tsx]: `import { clamp } from '../../utils/dom/dom';`

File: src/client/components/resize-handle/resize-handle.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    this.state = {};
+
+    this.onGlobalMouseUp = this.onGlobalMouseUp.bind(this);
+    this.onGlobalMouseMove = this.onGlobalMouseMove.bind(this);
+  }
+
+  componentDidMount() {
+    this.setState({
+      currentValue: this.props.initialValue
```
Comment (vogievetsky) [inline src/client/components/resize-handle/resize-handle.tsx]: you should constrain it here also

File: src/client/components/cube-view/cube-view.tsx [hunk 6]
```diff
@@ -41,6 +43,7 @@ export interface CubeViewState {
   showRawDataModal?: boolean;
   RawDataModalAsync?: typeof RawDataModal;
   downloadableDataset?: Dataset;
+  layout?: {left: number, right: number};
```
Comment (vogievetsky) [inline src/client/components/cube-view/cube-view.tsx]: I am not a huge fan of inline interfaces, could you move this into its own interface maybe call it `CubeViewLayout` ? Also I am not sure about 'left' and 'right', they sound too generic. Would a more descriptive name be better? Like `dimensionPanelWidth` and `pinboardWidth`

File: src/client/components/cube-view/cube-view.tsx [hunk 7]
```diff
@@ -301,11 +306,47 @@ export class CubeView extends React.Component<CubeViewProps, CubeViewState> {
     this.setState({ essence: newEsssence });
   }
 
+  getStoredLayout(): any {
+    let defaultLayout: any = {left: 240, right: 240};
+    let storedLayout: any = undefined;
+    if (localStorage.supported()) {
+      storedLayout = localStorage.get('cube-layout');
```
Comment (vogievetsky) [inline src/client/components/cube-view/cube-view.tsx]: could you change `cube-layout` to `cube-view-layout`, you'll see why soon. Muhahahaha

File: src/client/utils/local-storage/local-storage.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  if (value) {
+    try {
+      value = JSON.parse(value);
+      return value;
+    } catch (e) {
+      return undefined;
+    }
+  }
+
+  return undefined;
```
Comment (vogievetsky) [inline src/client/utils/local-storage/local-storage.ts]: I am pretty sure you could shorten this function to:

```
export function get(key: string): any {
  try {
    return JSON.parse(localStorage[key]);
  } catch(e) {
    return undefined;
  }
}
```

:-p

**Comment by vogievetsky** [comment]
Fixes: https://github.com/implydata/pivot/issues/109

**Comment by vogievetsky** [comment]
Also solves (3) in https://groups.google.com/forum/#!searchin/imply-user-group/pinboard/imply-user-group/Jz4diJwR0bo/enb8CEWhBAAJ

**Comment by vogievetsky** [comment]
gr 🎱

**Pull request merged**